In [3]:
# Libararies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Training Data
train_clinical = './train_clinical_data.csv'
train_peptides = './train_peptides.csv'
train_supp_clinical = './supplemental_clinical_data.csv'
train_proteins = './train_proteins.csv'

# Given Test Data
test_proteins= './example_test_files/test_proteins.csv'
test_peptides = './example_test_files/test_peptides.csv'
test_clinical = './example_test_files/test.csv'

## Load Training and Test Data

In [4]:
clinical_data = pd.read_csv(train_clinical)
protein_data = pd.read_csv(train_proteins)
peptide_data = pd.read_csv(train_peptides)
supp_data = pd.read_csv(train_supp_clinical)


clinical_data['upd23b_clinical_state_on_medication'] = clinical_data['upd23b_clinical_state_on_medication'].replace({'On': 1})
clinical_data['upd23b_clinical_state_on_medication'] = clinical_data['upd23b_clinical_state_on_medication'].replace({'Off': -1})
clinical_data['upd23b_clinical_state_on_medication'] = clinical_data['upd23b_clinical_state_on_medication'].fillna(0)

# Naming convention for me to use w/o ruining codyʻs code
train_proteins = protein_data
train_peptides = peptide_data
train_clinical = clinical_data

test_proteins = pd.read_csv(test_proteins)
test_peptides = pd.read_csv(test_peptides)
test_clinical = pd.read_csv(test_clinical)

## Imported Data

In [5]:
# Protein Training Data
print(train_proteins.shape)
train_proteins.head()

(232741, 5)


visit_id  visit_month  patient_id UniProt       NPX
0     55_0            0          55  O00391   11254.3
1     55_0            0          55  O00533  732430.0
2     55_0            0          55  O00584   39585.8
3     55_0            0          55  O14498   41526.9
4     55_0            0          55  O14773   31238.0

In [6]:
# Protein Testing Data
print(test_proteins.shape)
test_proteins.head()

(453, 6)


visit_id  visit_month  patient_id UniProt        NPX  group_key
0  50423_0            0       50423  O00391   33127.90          0
1  50423_0            0       50423  O00533  490742.00          0
2  50423_0            0       50423  O00584   43615.30          0
3  50423_0            0       50423  O14773   16486.60          0
4  50423_0            0       50423  O14791    2882.42          0

In [7]:
# Peptide Training Data
print(train_peptides.shape)
train_peptides.head()

(981834, 6)


visit_id  visit_month  patient_id UniProt  \
0     55_0            0          55  O00391   
1     55_0            0          55  O00533   
2     55_0            0          55  O00533   
3     55_0            0          55  O00533   
4     55_0            0          55  O00533   

                                  Peptide  PeptideAbundance  
0                           NEQEQPLGQWHLS           11254.3  
1                             GNPEPTFSWTK          102060.0  
2                         IEIPSSVQQVPTIIK          174185.0  
3  KPQSAVYSTGSNGILLC(UniMod_4)EAEGEPQPTIK           27278.9  
4                            SMEQNGPGLEYR           30838.7

In [8]:
# Peptide Testing Data
print(test_peptides.shape)
test_peptides.head()

(2057, 7)


visit_id  visit_month  patient_id UniProt  \
0  50423_0            0       50423  O00391   
1  50423_0            0       50423  O00391   
2  50423_0            0       50423  O00533   
3  50423_0            0       50423  O00533   
4  50423_0            0       50423  O00533   

                                  Peptide  PeptideAbundance  group_key  
0                    AHFSPSNIILDFPAAGSAAR           22226.3          0  
1                           NEQEQPLGQWHLS           10901.6          0  
2                             GNPEPTFSWTK           51499.4          0  
3                         IEIPSSVQQVPTIIK          125492.0          0  
4  KPQSAVYSTGSNGILLC(UniMod_4)EAEGEPQPTIK           23174.2          0

In [9]:
# CLinical Training Data
print(train_clinical.shape)
train_clinical.head()

(2615, 8)


visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3  updrs_4  \
0     55_0          55            0     10.0      6.0     15.0      NaN   
1     55_3          55            3     10.0      7.0     25.0      NaN   
2     55_6          55            6      8.0     10.0     34.0      NaN   
3     55_9          55            9      8.0      9.0     30.0      0.0   
4    55_12          55           12     10.0     10.0     41.0      0.0   

   upd23b_clinical_state_on_medication  
0                                  0.0  
1                                  0.0  
2                                  0.0  
3                                  1.0  
4                                  1.0

In [10]:
# CLinical Testing Data
print(test_clinical.shape)
test_clinical.head()

(16, 6)


visit_id  visit_month  patient_id updrs_test           row_id  group_key
0   3342_0            0        3342    updrs_1   3342_0_updrs_1          0
1   3342_0            0        3342    updrs_2   3342_0_updrs_2          0
2   3342_0            0        3342    updrs_3   3342_0_updrs_3          0
3   3342_0            0        3342    updrs_4   3342_0_updrs_4          0
4  50423_0            0       50423    updrs_1  50423_0_updrs_1          0

## Understanding The Data

In [11]:
train_proteins.groupby('visit_id')
train_proteins

visit_id  visit_month  patient_id UniProt       NPX
0            55_0            0          55  O00391   11254.3
1            55_0            0          55  O00533  732430.0
2            55_0            0          55  O00584   39585.8
3            55_0            0          55  O14498   41526.9
4            55_0            0          55  O14773   31238.0
...           ...          ...         ...     ...       ...
232736  58648_108          108       58648  Q9UBX5   27387.8
232737  58648_108          108       58648  Q9UHG2  369437.0
232738  58648_108          108       58648  Q9UKV8  105830.0
232739  58648_108          108       58648  Q9Y646   21257.6
232740  58648_108          108       58648  Q9Y6R7   17953.1

[232741 rows x 5 columns]

In [12]:
# Organize by visit_id (which includes the patient_id + visit_month)
# Protein data is aggregated based on information that is useful for me (# unique proteins, details about the NPX scores)
# Peptide data also include # unique peptides (multiple peptides make up a protein)
'''
Note: High NPX score means high protein concentration. If we find what proteins in parkinsons patients have high
concentrations, we can assume they are more correlated to the disease being present
'''
agg_protein_data = train_proteins.groupby('visit_id').agg({'UniProt':'nunique','NPX':['min','max','mean','std']}).reset_index()
agg_peptide_data = train_peptides.groupby('visit_id').agg({'UniProt':'nunique','Peptide':'nunique','PeptideAbundance': ['min','max','mean','std']}).reset_index()

In [13]:
agg_protein_data
# agg_protein_data.loc[agg_protein_data['visit_id'] == '55_0']

visit_id UniProt       NPX                                         
               nunique       min          max          mean           std
0      10053_0     165  2497.840  269126000.0  2.856580e+06  2.131630e+07
1     10053_12     171  5800.870  270030000.0  2.728871e+06  2.092162e+07
2     10053_18     208  1334.110  278835000.0  2.509967e+06  1.969453e+07
3     10138_12     217  2520.240  365582000.0  3.002583e+06  2.516170e+07
4     10138_24     219  1436.940  396894000.0  3.068891e+06  2.716806e+07
...        ...     ...       ...          ...           ...           ...
1108   8699_24     216   756.551  346067000.0  3.064059e+06  2.409420e+07
1109    942_12     212  1722.770  330558000.0  2.613298e+06  2.295228e+07
1110    942_24     217  1339.150  336769000.0  2.616142e+06  2.312662e+07
1111    942_48     216  1272.480  358059000.0  2.768442e+06  2.460543e+07
1112     942_6     216  2491.690  290111000.0  2.476880e+06  2.002174e+07

[1113 rows x 6 columns]

In [14]:
agg_peptide_data

visit_id UniProt Peptide PeptideAbundance                              \
               nunique nunique              min          max           mean   
0      10053_0     165     649          82.9679   66333900.0  726248.393431   
1     10053_12     171     633         128.4460   73059300.0  737183.385744   
2     10053_18     208     868         108.5000   64711200.0  601466.784320   
3     10138_12     217     932         129.0240   71652400.0  699099.199189   
4     10138_24     219     918         142.6480  123897000.0  732120.888877   
...        ...     ...     ...              ...          ...            ...   
1108   8699_24     216     911         106.9420   99846400.0  726494.824901   
1109    942_12     212     889          88.3277   70888500.0  623193.979635   
1110    942_24     217     910         108.7050   71995500.0  623849.652027   
1111    942_48     216     907         148.1360   70658500.0  659297.802601   
1112     942_6     216     918         158.8690   58472500.0  582795.558517   

                    
               std  
0     3.535602e+06  
1     3.799654e+06  
2     3.006568e+06  
3     3.379573e+06  
4     4.912602e+06  
...            ...  
1108  4.080307e+06  
1109  3.362987e+06  
1110  3.294163e+06  
1111  3.359265e+06  
1112  2.734257e+06  

[1113 rows x 7 columns]

Note that between peptide and protein data, the peptide frequency (peptideAbundance) and the protein frequency (essentially NPX), is the same for some UniProteins (1:1), but not all of them. This is because some proteins have multiple of the same peptide composing it. So some proteins have more peptideAbundance than NPX, which is why the agg() is not equivalent between the same patient visits with the same proteins.

## Reorganize Training DF

##### What Features Do We Need For Predictions
For easier use in the random forest regressor class, the training data is reorganized so that each row represents a sample patient record. Each column is some feature. The features we need are the visit_month, in order to sort, and the UniProtein ids so that we know the frequency of certain proteins in each patients visit. 

Lets look at the clinical test data once again ...

In [15]:
test_clinical

visit_id  visit_month  patient_id updrs_test           row_id  group_key
0    3342_0            0        3342    updrs_1   3342_0_updrs_1          0
1    3342_0            0        3342    updrs_2   3342_0_updrs_2          0
2    3342_0            0        3342    updrs_3   3342_0_updrs_3          0
3    3342_0            0        3342    updrs_4   3342_0_updrs_4          0
4   50423_0            0       50423    updrs_1  50423_0_updrs_1          0
5   50423_0            0       50423    updrs_2  50423_0_updrs_2          0
6   50423_0            0       50423    updrs_3  50423_0_updrs_3          0
7   50423_0            0       50423    updrs_4  50423_0_updrs_4          0
8    3342_6            6        3342    updrs_1   3342_6_updrs_1          6
9    3342_6            6        3342    updrs_2   3342_6_updrs_2          6
10   3342_6            6        3342    updrs_3   3342_6_updrs_3          6
11   3342_6            6        3342    updrs_4   3342_6_updrs_4          6
12  50423_6            6       50423    updrs_1  50423_6_updrs_1          6
13  50423_6            6       50423    updrs_2  50423_6_updrs_2          6
14  50423_6            6       50423    updrs_3  50423_6_updrs_3          6
15  50423_6            6       50423    updrs_4  50423_6_updrs_4          6

In [16]:
train_proteins

visit_id  visit_month  patient_id UniProt       NPX
0            55_0            0          55  O00391   11254.3
1            55_0            0          55  O00533  732430.0
2            55_0            0          55  O00584   39585.8
3            55_0            0          55  O14498   41526.9
4            55_0            0          55  O14773   31238.0
...           ...          ...         ...     ...       ...
232736  58648_108          108       58648  Q9UBX5   27387.8
232737  58648_108          108       58648  Q9UHG2  369437.0
232738  58648_108          108       58648  Q9UKV8  105830.0
232739  58648_108          108       58648  Q9Y646   21257.6
232740  58648_108          108       58648  Q9Y6R7   17953.1

[232741 rows x 5 columns]

We want to predict the updrs_1, updrs_2, updrs_3, and updrs_4 scores for patients 3342 and 50423 for months 0 and 6. 

In [17]:
# All we need is the patient, month, and the NPX scores of all the UniProteins
def create_prot_df(prot_df):
    records = []   # Stores the list of features for each training sample (each row of the df)
    for grp_idx, data in prot_df:    # grp_idx = grouped index tuple of (patient_id, visit_month), data = a row of the df
        feats = list(grp_idx)
        # For each row, we check a list of all proteins. If any protein is in that sample row (data), we add it that rows list of features
        # Else, we add 0 (that protein doesn't exist for this patient_id)
        for j in range(len(all_prot)):
            if all_prot[j] in data['UniProt'].values: 
                feats.append(data[data.UniProt == all_prot[j]].NPX.iloc[0])
            else:
                feats.append(0.0)
        records.append(feats)
    # Create the dataframe w/ patient, month, and NPX scores of all proteins that each patient_id has > 0
    df = pd.DataFrame(records, columns=['patient_id', 'visit_month'] + all_prot)
    return df

In [18]:
# Sorted list of uniprot ids
all_prot = sorted(set(train_proteins.UniProt).union(set(train_peptides.UniProt)))  # 227 proteins total
# Group protein data by patient_id and month
prot_grp = train_proteins.groupby(["patient_id", "visit_month"])

In [19]:
# Create the protein feature df
df_npx = create_prot_df(prot_grp)
df_npx

patient_id  visit_month   O00391     O00533   O00584   O14498    O14773  \
0             55            0  11254.3   732430.0  39585.8  41526.9  31238.00   
1             55            6  13163.6   630465.0  35220.8  41295.0  26219.90   
2             55           12  15257.6   815083.0  41650.9  39763.3  30703.60   
3             55           36  13530.8   753832.0  43048.9  43503.6  33577.60   
4            942            6  11218.7   399518.0  20581.0  31290.9   6173.58   
...          ...          ...      ...        ...      ...      ...       ...   
1108       64674           84      0.0   190487.0  24907.9  18543.1  10124.90   
1109       65043            0  13472.4   927954.0  42661.5  43663.2  20071.30   
1110       65043           12  14134.9   984651.0  28990.8  42440.9  25357.40   
1111       65043           24  14659.5  1062020.0  46440.4  38293.0  21971.80   
1112       65043           48  10589.6   902434.0  44890.8  38771.5  20188.30   

       O14791    O15240    O15394  ...    Q9HDC9    Q9NQ79    Q9NYU2   Q9UBR2  \
0     4202.71  177775.0   62898.2  ...  365475.0  35528.00   97005.6  23122.5   
1     4416.42  165638.0   62567.5  ...  405676.0  30332.60  109174.0  23499.8   
2     4343.60  151073.0   66963.1  ...  303953.0  43026.20  114921.0  21860.1   
3     5367.06  101056.0   67588.6  ...  303597.0  48188.40  109794.0  23930.6   
4     2564.37  160526.0   43423.1  ...  253373.0  27431.80   93796.7  17450.9   
...       ...       ...       ...  ...       ...       ...       ...      ...   
1108  2308.71   62095.4   29248.7  ...  260021.0   7139.93  104277.0  10500.0   
1109  3278.88  266339.0  117884.0  ...  186414.0  25897.80       0.0  21480.7   
1110  3267.66  270575.0  108246.0  ...  301343.0  22343.40  105626.0  20500.8   
1111  3990.34  221358.0  117204.0  ...  300439.0  52143.60  139291.0  19449.2   
1112  3137.07  203487.0   85782.0  ...  223136.0  21659.90  116470.0  14073.3   

       Q9UBX5    Q9UHG2   Q9UKV8   Q9UNU6   Q9Y646   Q9Y6R7  
0     60912.6  408698.0      0.0  29758.8  23833.7  18953.5  
1     51655.8  369870.0      0.0  22935.2  17722.5  16642.7  
2     61598.2  318553.0  65762.6  29193.4  28536.1  19290.9  
3     70223.5  377550.0  74976.1  31732.6  22186.5  21717.1  
4     21299.1  306621.0  82335.5  24018.7  18939.5  15251.2  
...       ...       ...      ...      ...      ...      ...  
1108  21944.2  136725.0  62217.5      0.0  10287.7  13848.2  
1109  57364.0  416142.0  37584.6      0.0  28346.5  35617.5  
1110  54011.2  380072.0  40588.9      0.0  17035.7  37064.2  
1111  66569.9  300948.0  36150.4      0.0  21286.3  39587.9  
1112  48796.4  320821.0  39046.7      0.0  20198.8  39535.0  

[1113 rows x 229 columns]

In [20]:
'''
Calculate the freq of each UniProt by taking the mean of all the non-zero NPX values for each protein ID. Each protein is mapped
to be either 0 or 1, indicating if it is present or not in the specific patient_id + month. Then we take the mean of this
mapping, such that we are getting a frequency between [0, 1] on how often the protein is present across all the patients. 
Ideally, the proteins that aren't present very often can be ignored as they probably don't have much to do with the prediction
of updrs scores.
'''

# Actually I dont want to have a frequency cutoff because I already lost to many training samples
FREQ_CUTOFF = 0.0
prot_freq = (df_npx.iloc[:, 2:] > 0).mean(axis = 0)   
print(prot_freq)
prot_select = prot_freq.index[(prot_freq >= FREQ_CUTOFF)].tolist()
# now merge in the UPDRS data from the clinical dataset
# we only merge data between train_clinical and train_proteins if they share the same patient_id + visit_month
clin_cols = ["patient_id", "visit_month", "updrs_1", "updrs_2", "updrs_3", "updrs_4"]
join = df_npx[["patient_id", "visit_month"] + prot_select]\
    .merge(train_clinical[clin_cols], on=["patient_id", "visit_month"])
join.head()

O00391    0.686433
O00533    0.999102
O00584    0.988320
O14498    0.927224
O14773    0.940701
            ...   
Q9UHG2    1.000000
Q9UKV8    0.814915
Q9UNU6    0.683738
Q9Y646    0.945193
Q9Y6R7    0.938005
Length: 227, dtype: float64


patient_id  visit_month   O00391    O00533   O00584   O14498    O14773  \
0          55            0  11254.3  732430.0  39585.8  41526.9  31238.00   
1          55            6  13163.6  630465.0  35220.8  41295.0  26219.90   
2          55           12  15257.6  815083.0  41650.9  39763.3  30703.60   
3          55           36  13530.8  753832.0  43048.9  43503.6  33577.60   
4         942            6  11218.7  399518.0  20581.0  31290.9   6173.58   

    O14791    O15240   O15394  ...   Q9UBX5    Q9UHG2   Q9UKV8   Q9UNU6  \
0  4202.71  177775.0  62898.2  ...  60912.6  408698.0      0.0  29758.8   
1  4416.42  165638.0  62567.5  ...  51655.8  369870.0      0.0  22935.2   
2  4343.60  151073.0  66963.1  ...  61598.2  318553.0  65762.6  29193.4   
3  5367.06  101056.0  67588.6  ...  70223.5  377550.0  74976.1  31732.6   
4  2564.37  160526.0  43423.1  ...  21299.1  306621.0  82335.5  24018.7   

    Q9Y646   Q9Y6R7  updrs_1  updrs_2  updrs_3  updrs_4  
0  23833.7  18953.5     10.0      6.0     15.0      NaN  
1  17722.5  16642.7      8.0     10.0     34.0      NaN  
2  28536.1  19290.9     10.0     10.0     41.0      0.0  
3  22186.5  21717.1     17.0     18.0     51.0      0.0  
4  18939.5  15251.2      8.0      2.0     21.0      NaN  

[5 rows x 233 columns]

In [21]:
join

patient_id  visit_month   O00391     O00533   O00584   O14498    O14773  \
0             55            0  11254.3   732430.0  39585.8  41526.9  31238.00   
1             55            6  13163.6   630465.0  35220.8  41295.0  26219.90   
2             55           12  15257.6   815083.0  41650.9  39763.3  30703.60   
3             55           36  13530.8   753832.0  43048.9  43503.6  33577.60   
4            942            6  11218.7   399518.0  20581.0  31290.9   6173.58   
...          ...          ...      ...        ...      ...      ...       ...   
1063       64674           84      0.0   190487.0  24907.9  18543.1  10124.90   
1064       65043            0  13472.4   927954.0  42661.5  43663.2  20071.30   
1065       65043           12  14134.9   984651.0  28990.8  42440.9  25357.40   
1066       65043           24  14659.5  1062020.0  46440.4  38293.0  21971.80   
1067       65043           48  10589.6   902434.0  44890.8  38771.5  20188.30   

       O14791    O15240    O15394  ...   Q9UBX5    Q9UHG2   Q9UKV8   Q9UNU6  \
0     4202.71  177775.0   62898.2  ...  60912.6  408698.0      0.0  29758.8   
1     4416.42  165638.0   62567.5  ...  51655.8  369870.0      0.0  22935.2   
2     4343.60  151073.0   66963.1  ...  61598.2  318553.0  65762.6  29193.4   
3     5367.06  101056.0   67588.6  ...  70223.5  377550.0  74976.1  31732.6   
4     2564.37  160526.0   43423.1  ...  21299.1  306621.0  82335.5  24018.7   
...       ...       ...       ...  ...      ...       ...      ...      ...   
1063  2308.71   62095.4   29248.7  ...  21944.2  136725.0  62217.5      0.0   
1064  3278.88  266339.0  117884.0  ...  57364.0  416142.0  37584.6      0.0   
1065  3267.66  270575.0  108246.0  ...  54011.2  380072.0  40588.9      0.0   
1066  3990.34  221358.0  117204.0  ...  66569.9  300948.0  36150.4      0.0   
1067  3137.07  203487.0   85782.0  ...  48796.4  320821.0  39046.7      0.0   

       Q9Y646   Q9Y6R7  updrs_1  updrs_2  updrs_3  updrs_4  
0     23833.7  18953.5     10.0      6.0     15.0      NaN  
1     17722.5  16642.7      8.0     10.0     34.0      NaN  
2     28536.1  19290.9     10.0     10.0     41.0      0.0  
3     22186.5  21717.1     17.0     18.0     51.0      0.0  
4     18939.5  15251.2      8.0      2.0     21.0      NaN  
...       ...      ...      ...      ...      ...      ...  
1063  10287.7  13848.2     11.0     15.0     45.0      4.0  
1064  28346.5  35617.5      2.0      6.0     16.0      NaN  
1065  17035.7  37064.2      4.0      7.0     14.0      0.0  
1066  21286.3  39587.9      4.0      8.0      NaN      0.0  
1067  20198.8  39535.0      7.0      6.0     13.0      0.0  

[1068 rows x 233 columns]

To predict month 0 we can use both month 0 and 3. 

In [22]:
df_proteins_0 = join.loc[(join['visit_month'] == 0)]
df_proteins_0 = df_proteins_0.drop('updrs_4', axis=1)
df_proteins_6 = join.loc[(join['visit_month'] == 6)]
df_proteins_6 = df_proteins_6.drop('updrs_4', axis=1)
common_rows = pd.merge(df_proteins_0, df_proteins_6, on='patient_id')
common_rows

patient_id  visit_month_x  O00391_x   O00533_x  O00584_x  O14498_x  \
0           55              0  11254.30   732430.0   39585.8   41526.9   
1         4161              0  12336.20  1074750.0   25027.3   52936.4   
2         5645              0   9846.73   539888.0   36067.8   26773.5   
3         5742              0   7405.52   223579.0   19316.2   27151.9   
4         6211              0      0.00   187035.0   18411.9   19501.2   
..         ...            ...       ...        ...       ...       ...   
63       61503              0  11248.40   710658.0   28116.3   24842.7   
64       62723              0      0.00   466328.0   26596.2   19756.0   
65       62792              0  15119.20  1038820.0   42199.0   39770.3   
66       64669              0  10806.50   491365.0   26262.8   26712.7   
67       64674              0      0.00   351914.0   21625.1   17712.4   

    O14773_x  O14791_x  O15240_x  O15394_x  ...  Q9UBR2_y  Q9UBX5_y  Q9UHG2_y  \
0   31238.00   4202.71  177775.0   62898.2  ...  23499.80   51655.8  369870.0   
1       0.00   2180.17  249566.0   45125.8  ...  11221.90   23230.0  365088.0   
2   43874.00      0.00  174718.0   64748.3  ...  23973.80   31343.8  242057.0   
3       0.00   2550.78   91819.3       0.0  ...   9302.28   26005.7  227630.0   
4    4420.69   1627.70   39372.2   13175.4  ...   7600.02   23326.5  158637.0   
..       ...       ...       ...       ...  ...       ...       ...       ...   
63  20251.90   2004.91  157261.0   43692.1  ...  15763.10   32893.5  188276.0   
64   7752.63   2649.08  107580.0   41415.9  ...  13273.40   32026.6  230125.0   
65  20818.00   3021.31  286860.0   98972.1  ...  21313.90   49049.3  383056.0   
66  15372.50   2995.08   73504.6   57807.2  ...  13971.40   35030.0  144959.0   
67  20849.80      0.00   69664.3   31173.7  ...   6771.70   29385.9  100466.0   

    Q9UKV8_y  Q9UNU6_y  Q9Y646_y  Q9Y6R7_y  updrs_1_y  updrs_2_y  updrs_3_y  
0        0.0   22935.2  17722.50   16642.7        8.0       10.0       34.0  
1        0.0       0.0      0.00   23840.1        1.0        2.0        0.0  
2    34444.9   27677.6  30311.80    9024.8        7.0       11.0       25.0  
3    46723.3   18423.8   9144.79   10893.1        6.0       13.0       56.0  
4        0.0       0.0   9283.77   11768.7        6.0       11.0       21.0  
..       ...       ...       ...       ...        ...        ...        ...  
63   73591.8   15222.4  19682.00   16264.1        1.0        7.0       27.0  
64   67565.9       0.0  17826.10   14096.9        8.0        2.0        8.0  
65       0.0       0.0  21886.90   45358.2        0.0        3.0        7.0  
66   58798.5       0.0  17526.60   19882.2       17.0       20.0       33.0  
67   80147.2   14217.5  22023.60   11517.0       11.0        7.0       20.0  

[68 rows x 463 columns]

In [23]:
# Split the common_rows df back into months 0 and months 6

# Get visit_id column
visit_id_df = common_rows.iloc[:,0]

# Split the common_rows back to month 0 and 6 respectively
df_proteins_0 = common_rows.loc[:, common_rows.columns.str.endswith('x')]
df_proteins_6 = common_rows.loc[:, common_rows.columns.str.endswith('y')]

# Merge visit_id back to each proteins dataframe
# Also strip the trailing strings from the column labels
df_proteins_0 = pd.concat([visit_id_df, df_proteins_0], axis=1)
df_proteins_6 = pd.concat([visit_id_df, df_proteins_6], axis=1)
df_proteins_0.columns = df_proteins_0.columns.str.rstrip('_x')
df_proteins_6.columns = df_proteins_6.columns.str.rstrip('_y')

# Shared visit id of month 0
df_proteins_0

patient_id  visit_month    O00391     O00533   O00584   O14498    O14773  \
0           55            0  11254.30   732430.0  39585.8  41526.9  31238.00   
1         4161            0  12336.20  1074750.0  25027.3  52936.4      0.00   
2         5645            0   9846.73   539888.0  36067.8  26773.5  43874.00   
3         5742            0   7405.52   223579.0  19316.2  27151.9      0.00   
4         6211            0      0.00   187035.0  18411.9  19501.2   4420.69   
..         ...          ...       ...        ...      ...      ...       ...   
63       61503            0  11248.40   710658.0  28116.3  24842.7  20251.90   
64       62723            0      0.00   466328.0  26596.2  19756.0   7752.63   
65       62792            0  15119.20  1038820.0  42199.0  39770.3  20818.00   
66       64669            0  10806.50   491365.0  26262.8  26712.7  15372.50   
67       64674            0      0.00   351914.0  21625.1  17712.4  20849.80   

     O14791    O15240   O15394  ...    Q9UBR2   Q9UBX5    Q9UHG2   Q9UKV8  \
0   4202.71  177775.0  62898.2  ...  23122.50  60912.6  408698.0      0.0   
1   2180.17  249566.0  45125.8  ...   9602.96  18350.4  543219.0  67290.7   
2      0.00  174718.0  64748.3  ...  21376.60  36594.1  315995.0  31557.5   
3   2550.78   91819.3      0.0  ...   8371.27  24024.0  169177.0  53700.1   
4   1627.70   39372.2  13175.4  ...      0.00  25247.7  164970.0  35214.0   
..      ...       ...      ...  ...       ...      ...       ...      ...   
63  2004.91  157261.0  43692.1  ...  11701.00  33642.4  202296.0      0.0   
64  2649.08  107580.0  41415.9  ...  16454.40  31906.4  177486.0  69520.8   
65  3021.31  286860.0  98972.1  ...  20480.80  48041.4  394514.0  70445.0   
66  2995.08   73504.6  57807.2  ...  15743.50  27776.3  171664.0  63783.2   
67     0.00   69664.3  31173.7  ...  10139.90  28097.8  150243.0  64857.7   

     Q9UNU6    Q9Y646    Q9Y6R7  updrs_1  updrs_2  updrs_3  
0   29758.8  23833.70  18953.50     10.0      6.0     15.0  
1       0.0  13029.40  35091.00      6.0      1.0      0.0  
2   25243.0  36819.80   9909.75     12.0      7.0     17.0  
3   16306.8   5454.99   8131.81      5.0      7.0     38.0  
4       0.0  13859.90  13141.60      5.0     11.0      9.0  
..      ...       ...       ...      ...      ...      ...  
63  16702.4  24882.00  18060.80      3.0      5.0     20.0  
64  11264.7  20154.30  14874.10      6.0      2.0      4.0  
65      0.0  17492.70  29112.00      0.0      2.0     10.0  
66      0.0      0.00  25094.20     12.0     14.0     27.0  
67  14324.6  22257.60  10510.50      5.0      1.0     13.0  

[68 rows x 232 columns]

In [24]:
# Shared visit id of month 6
df_proteins_6

patient_id  visit_month   O00391    O00533   O00584   O14498    O14773  \
0           55            6  13163.6  630465.0  35220.8  41295.0  26219.90   
1         4161            6  11191.5  836088.0  18232.4  44420.0      0.00   
2         5645            6  13070.0  608796.0  34771.8      0.0  32705.30   
3         5742            6      0.0  398967.0  23080.5  30432.5      0.00   
4         6211            6      0.0  183667.0  13686.4  16421.7   6851.54   
..         ...          ...      ...       ...      ...      ...       ...   
63       61503            6  11674.3  724274.0  24918.9  26372.7  21431.70   
64       62723            6      0.0  446637.0  25861.0  25126.2  19681.20   
65       62792            6  13772.0  771691.0  41596.5  39963.9  23437.20   
66       64669            6      0.0  364638.0  23734.6  20410.8  12074.70   
67       64674            6      0.0  274549.0  21607.1  13334.7  15174.40   

     O14791    O15240   O15394  ...    Q9UBR2   Q9UBX5    Q9UHG2   Q9UKV8  \
0   4416.42  165638.0  62567.5  ...  23499.80  51655.8  369870.0      0.0   
1   3727.25  213213.0  31698.5  ...  11221.90  23230.0  365088.0      0.0   
2   2717.30  185609.0  39250.5  ...  23973.80  31343.8  242057.0  34444.9   
3   1006.44  126966.0      0.0  ...   9302.28  26005.7  227630.0  46723.3   
4   2188.71   44039.4  12870.0  ...   7600.02  23326.5  158637.0      0.0   
..      ...       ...      ...  ...       ...      ...       ...      ...   
63   924.61  165806.0  72638.7  ...  15763.10  32893.5  188276.0  73591.8   
64  3362.51  125738.0  40699.2  ...  13273.40  32026.6  230125.0  67565.9   
65  2160.20  237267.0  87903.0  ...  21313.90  49049.3  383056.0      0.0   
66  2864.83   70720.8  35008.5  ...  13971.40  35030.0  144959.0  58798.5   
67  2240.71   72555.1  27139.5  ...   6771.70  29385.9  100466.0  80147.2   

     Q9UNU6    Q9Y646   Q9Y6R7  updrs_1  updrs_2  updrs_3  
0   22935.2  17722.50  16642.7      8.0     10.0     34.0  
1       0.0      0.00  23840.1      1.0      2.0      0.0  
2   27677.6  30311.80   9024.8      7.0     11.0     25.0  
3   18423.8   9144.79  10893.1      6.0     13.0     56.0  
4       0.0   9283.77  11768.7      6.0     11.0     21.0  
..      ...       ...      ...      ...      ...      ...  
63  15222.4  19682.00  16264.1      1.0      7.0     27.0  
64      0.0  17826.10  14096.9      8.0      2.0      8.0  
65      0.0  21886.90  45358.2      0.0      3.0      7.0  
66      0.0  17526.60  19882.2     17.0     20.0     33.0  
67  14217.5  22023.60  11517.0     11.0      7.0     20.0  

[68 rows x 232 columns]

## Training DataFrame
df_proteins_0 is the common visit id's from month 0 and df_proteins_6 are the same patient visits in month 6. Therefore, I will train the model with the month 0 protein npx scores and predict the updrs on month 6. I need to take the protein data from df_proteins_0 and split it to test and train data. I will do this manually as to not test updrs scores of mixed patient id's. I will also repeat this with the updrs columns in df_protein_6 which will be y_train and y_test.

In [25]:
### 68 rows means 55 training rows and 13 test rows for a  0.8:0.2 train test split.
X_train = df_proteins_0.sample(n=55, random_state=15)
X_test = df_proteins_0.drop(X_train.index)
y_train = df_proteins_0.sample(n=55, random_state=15)
y_test = df_proteins_0.drop(y_train.index)

# Check if the split worked
print(len(X_train))
print(len(X_test))
print(len(y_train))
print(len(y_test))

if X_train['patient_id'].equals(y_train['patient_id']): # Is the rows of the features X and targets y in the same order
    print("Train rows are the same order")
if X_test['patient_id'].equals(y_test['patient_id']):
    print("Test rows are the same order")
    
# Grab the columns we want
X_train = X_train.iloc[:,2:-3]
X_test = X_test.iloc[:,2:-3]
y_train = y_train.iloc[:,-3:]
y_test = y_test.iloc[:,-3:]

55
13
55
13
Train rows are the same order
Test rows are the same order


At this point X_train, y_train, X_test, and y_test should have the features needed to train. A lot of samples have been lost during the dataframe cleanup so the results may not be very good.

In [26]:
# Mean updrs scores (1 through 4) for each month
updrs_means = train_clinical.groupby("visit_month")[["updrs_1", "updrs_2", "updrs_3", "updrs_4"]].mean()
updrs_means

updrs_1   updrs_2    updrs_3   updrs_4
visit_month                                         
0            5.572581  4.439516  13.754032  2.047619
3            5.513043  6.634783  20.482456  0.142857
6            7.109375  6.963542  20.272251  2.307692
9            6.080808  7.204082  20.474747  0.488372
12           6.222222  5.320988  16.165975  1.308271
18           7.299465  6.866310  19.016216  1.218978
24           6.670782  5.658436  16.489540  1.658065
30           8.231214  7.647399  21.710983  1.753425
36           7.331858  6.230088  18.316964  1.765432
42           8.339869  8.372549  22.720779  1.868613
48           7.581633  7.112245  19.378238  1.970588
54           8.600000  9.863636  26.311927  2.051020
60           7.265060  6.500000  19.807229  2.218182
72           8.365591  8.817204  26.450549  2.289157
84           7.730000  8.230000  22.589474  2.938462
96           7.706897  7.775862  21.267857  4.233333
108          9.416667  8.166667  25.250000  2.571429

## Import Model Libraries

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split, StratifiedKFold, GridSearchCV, cross_val_score
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import confusion_matrix, mean_squared_error
from sklearn.metrics import accuracy_score
import math

## Random Forest Classifier vs Bagging Classifier Background Info
Both ensemble methods utilize multiple decision trees to perform classification. However, they differ in a few ways. 

#### Random Forest
In random forest classification, a random subset of features is selected for each iteration of the test_train_split. Random forests sample without replacement, meaning each sample instance can be selected only once per sample. Finally, when the training is done, the majority vote is based on a subset of randomly selected decision trees, rather than all the decision trees. This is done to reduce overfitting on the training data.

#### Bagging
On the other hand, bagging is slightly less robust than random forest classification. Instead of generalizing the model, in bagging, sample instances are selected with replacement. This means that there is a chance that a training sample can have instances of the same data. E.g. Given a set {0, 1, 2}, in random forest classifiers, a sample has to have unique instances, like {1, 2, 0}, {0, 2, 1}, etc. But in bagging, there is replacement so we could have {0, 0, 1} or event {2, 2, 2}. In addition, when we reach the end of training, the final decision is based on the output of every decision tree, which may cause overfitting.

## Random Forest Regressor Training

In [28]:
X_train.head()

O00391    O00533   O00584   O14498    O14773   O14791    O15240  \
18      0.00  328743.0  38082.3  31337.3  24477.80  1745.21   61084.5   
33   9542.65  394154.0  29288.2  33274.2  12731.50  2050.77  141926.0   
55      0.00  630834.0  13170.0  28752.7   9341.57  1134.19  220760.0   
64      0.00  466328.0  26596.2  19756.0   7752.63  2649.08  107580.0   
30  10596.50  607245.0  21681.1  29530.2  17977.10  3014.10  145501.0   

     O15394    O43505    O60888  ...    Q9HDC9   Q9NQ79    Q9NYU2   Q9UBR2  \
18  43429.0  198892.0  117129.0  ...  200924.0  23546.4   87674.6  17509.8   
33      0.0  185152.0  145775.0  ...  246918.0  36193.7  113799.0  12644.3   
55  48545.9  361844.0  233741.0  ...       0.0  59928.1   94994.8  10795.4   
64  41415.9  243750.0  112355.0  ...  435201.0  56646.2  188308.0  16454.4   
30  65107.5  220374.0  142993.0  ...  327704.0  39985.0  104904.0  15923.8   

     Q9UBX5    Q9UHG2   Q9UKV8   Q9UNU6   Q9Y646    Q9Y6R7  
18  34873.7  222020.0  73452.0  26036.0  20439.2  34178.10  
33  33224.3  229211.0  29936.8  25299.3  19773.6  12669.80  
55  36413.7  271416.0  57248.4      0.0  16079.6   8551.65  
64  31906.4  177486.0  69520.8  11264.7  20154.3  14874.10  
30  24100.0  290666.0  78080.5      0.0  18620.7      0.00  

[5 rows x 227 columns]

In [29]:
X_test.head()

O00391    O00533   O00584   O14498    O14773    O14791    O15240  \
0   11254.3  732430.0  39585.8  41526.9  31238.00  4202.710  177775.0   
5       0.0  378596.0  34720.7  24792.9      0.00  2739.090  119623.0   
7       0.0  316400.0  20988.5  17683.7   9126.46  3038.810   86087.3   
11  12016.3  758936.0  32504.4  24892.4  15005.00  1947.850  152720.0   
12  10507.7  642774.0  33806.6  24485.6  19584.20   942.321  165562.0   

     O15394    O43505    O60888  ...    Q9HDC9    Q9NQ79    Q9NYU2    Q9UBR2  \
0   62898.2  333376.0  166850.0  ...  365475.0  35528.00   97005.6  23122.50   
5   47468.1  190576.0  117932.0  ...  364011.0  30752.00   83825.1      0.00   
7   35985.9  115798.0  134768.0  ...  168052.0   8181.95   85578.5   6919.08   
11  57580.4  241071.0  211649.0  ...   46756.0  15021.30  115206.0   8601.45   
12  50264.2  199377.0  206112.0  ...  283839.0  48947.60   84216.4  14990.80   

     Q9UBX5    Q9UHG2   Q9UKV8   Q9UNU6   Q9Y646   Q9Y6R7  
0   60912.6  408698.0      0.0  29758.8  23833.7  18953.5  
5   22168.2  231833.0  69985.5  21851.9  17868.5  26167.9  
7   22186.7  139896.0  38764.6  12248.1  16193.7  11248.2  
11  46818.5  288143.0  54758.9  15013.0  32860.0  12280.6  
12  61374.4  274244.0  96204.8  24447.6  25467.3  16798.2  

[5 rows x 227 columns]

In [30]:
# Lets look at all the training targets for updrs 1-3
y_train

updrs_1  updrs_2  updrs_3
18      1.0      6.0     46.0
33      3.0      2.0     20.0
55      5.0      6.0     22.0
64      6.0      2.0      4.0
30      6.0      6.0     26.0
42      3.0      7.0     20.0
67      5.0      1.0     13.0
53      1.0      2.0     12.0
43      3.0      4.0     22.0
56      0.0      3.0     31.0
36      8.0      2.0     16.0
14      4.0      4.0     18.0
16      1.0      1.0     12.0
24      1.0      1.0     23.0
25      7.0     22.0     41.0
20     10.0      2.0     13.0
66     12.0     14.0     27.0
65      0.0      2.0     10.0
51     13.0      9.0     19.0
63      3.0      5.0     20.0
62      1.0      7.0     19.0
49      9.0      6.0     24.0
3       5.0      7.0     38.0
58      9.0      2.0     23.0
34      1.0      4.0      7.0
22     15.0      9.0     17.0
46      2.0      7.0     26.0
2      12.0      7.0     17.0
9       6.0      3.0     16.0
6       2.0      2.0     13.0
35      3.0      0.0      5.0
8       2.0      2.0     14.0
40     18.0     13.0     23.0
61     10.0      6.0     26.0
26      2.0      2.0     10.0
60      1.0      7.0     28.0
52      2.0      1.0      1.0
59      6.0      3.0     15.0
1       6.0      1.0      0.0
48      2.0      4.0     10.0
38     18.0     13.0     39.0
50      5.0      5.0     11.0
57      4.0      4.0     26.0
13      7.0      2.0      5.0
44      1.0      2.0      7.0
19      1.0      2.0     14.0
37      5.0     14.0     25.0
39      1.0      2.0      8.0
41     15.0      9.0     29.0
4       5.0     11.0      9.0
15      1.0      2.0     15.0
10     12.0     20.0     36.0
32      8.0      5.0     12.0
23      0.0      3.0     13.0
31      8.0      7.0     21.0

In [32]:
# Lets look at all the test targets for updrs 1-3
y_test

updrs_1  updrs_2  updrs_3
0      10.0      6.0     15.0
5       4.0      1.0     16.0
7       2.0      6.0     21.0
11      5.0     15.0     22.0
12      8.0     11.0     22.0
17      1.0      3.0     20.0
21      5.0      5.0     23.0
27      3.0      3.0     15.0
28      6.0     15.0     18.0
29      9.0      8.0     20.0
45      6.0      0.0      2.0
47      1.0      2.0     14.0
54     14.0     15.0     26.0

In order to break up the target matrices into classes based on the percentile of the updrs scores, we first find the max and min values in each column. Then, we quantify the percentile ranges of it by [0.25, 0.50, 0.75, 1]. E.g. for scores between 0-10, find  the difference between the max and min -> 10-0=10. Then take the percentiles [10*0.25, 10*0.5, 10*0.75, 10*1] = [2.5, 5, 7.5, 10]. Any updrs_1 scores that fall in between these scores get placed in class [0, 1, 2, 3]. The borders of each percentile is inclusive. In the end the updrs matrices should only have values of 0 - 3. Then when we predict we classify from 0-3. And when we have accuracy, we are predicting what percentile of updrs scores the patient will likely be in for month 6, given there scores from month 0.

In [ ]:
# Find max and min of columns
for col in 

#### Validate the dataframes are properly synced

In [48]:
# X_train and y_train should have same matrix # rows. The same for _test
if X_train.shape[0] == y_train.shape[0]:
    print("Train Data: Success")
if X_test.shape[0] == y_test.shape[0]:
    print("Test Data: Success")

Train Data: Success
Test Data: Success


Try with a random forest classifier. Quantize the updrs scores into 4 percintiles and classify the output instead of a continuous regression.
When the r^2 score is negative, we are just predicting the same value all the time. So that if you guess the correct number, the accuracy will be higher. 

## UPDRS 1

In [58]:
y_train_1 = y_train['updrs_1']
y_test_1 = y_test['updrs_1']
model1 = RandomForestRegressor(n_estimators = 100, max_features="sqrt", min_samples_split=10)
model1.fit(X_train, y_train_1)
preds = model1.predict(X_test)

# compute the accuracy of the model
from sklearn.metrics import mean_squared_error, r2_score

# compute the mean squared error of the model
mse = mean_squared_error(y_test_1, preds)
print("MSE:", mse)

# compute the R2 score of the model
r2 = r2_score(y_test_1, preds)
print("R2:", r2)

# Cross Validation Score
scores = cross_val_score(model1, X_train, y_train_1, cv=5)
print("%0.2f accuracy with a standard deviation of %0.2f" % (scores.mean(), scores.std()))

MSE: 15.623299450140381
R2: -0.17557328899097246
-0.17 accuracy with a standard deviation of 0.26


In [50]:
print(preds.reshape(-1,1))
y_test_1

[[7.09291604]
 [5.57935386]
 [5.83793978]
 [4.10599893]
 [4.45287736]
 [8.4789482 ]
 [6.52198598]
 [4.85658639]
 [6.06613357]
 [5.4991669 ]
 [5.9009671 ]
 [6.29144139]
 [5.20756043]]


0     10.0
5      4.0
7      2.0
11     5.0
12     8.0
17     1.0
21     5.0
27     3.0
28     6.0
29     9.0
45     6.0
47     1.0
54    14.0
Name: updrs_1, dtype: float64

## UPDRS 2

In [51]:
y_train_2 = y_train['updrs_2']
y_test_2 = y_test['updrs_2']
model2 = RandomForestRegressor(n_estimators = 300)
model2.fit(X_train, y_train_2)
preds = model2.predict(X_test)

# compute the accuracy of the model
from sklearn.metrics import mean_squared_error, r2_score

# compute the mean squared error of the model
mse = mean_squared_error(y_test_2, preds)
print("MSE:", mse)

# compute the R2 score of the model
r2 = r2_score(y_test_2, preds)
print("R2:", r2)

# Cross Validation Score
scores = cross_val_score(model2, X_train, y_train_2, cv=5)
print("%0.2f accuracy with a standard deviation of %0.2f" % (scores.mean(), scores.std()))

MSE: 34.79652649572649
R2: -0.2673734865900379
-0.53 accuracy with a standard deviation of 0.57


## UPDRS 3

In [52]:
y_train_3 = y_train['updrs_3']
y_test_3 = y_test['updrs_3']
model3 = RandomForestRegressor(n_estimators = 300)
model3.fit(X_train, y_train_3)
preds = model3.predict(X_test)

# compute the accuracy of the model
from sklearn.metrics import mean_squared_error, r2_score

# compute the mean squared error of the model
mse = mean_squared_error(y_test_3, preds)
print("MSE:", mse)

# compute the R2 score of the model
r2 = r2_score(y_test_3, preds)
print("R2:", r2)

# Cross Validation Score
scores = cross_val_score(model3, X_train, y_train_3, cv=5)
print("%0.2f accuracy with a standard deviation of %0.2f" % (scores.mean(), scores.std()))

MSE: 42.44438803418803
R2: -0.27726167695473247
-0.19 accuracy with a standard deviation of 0.16
